### PR TITLE
Add getConnectionForSql and use it

### PR DESCRIPTION
**Pull request by srknzl**
Followup to #963. In 5.0 SQL team decided to use their own logic for selecting a connection for sending SQL message instead of using LoadBalancer. 

Like java side did, this pr deprecates old methods on load balancer.

### CODE REVIEW

File: src/BuildInfo.ts [hunk 1]
```diff
@@ -52,4 +53,7 @@ export class BuildInfo {
         return clientVersion;
     }
 
+    public static calculateMemberVersion(m: MemberVersion) : number {
+        return BuildInfo.calculateServerVersion(m.major, m.minor, m.patch);
+    }
```
Comment (srknzl) [inline src/BuildInfo.ts:58]: This is added as a utility

File: src/HazelcastClient.ts [hunk 2]
```diff
@@ -161,7 +161,8 @@ export class HazelcastClient {
             this.clusterFailoverService
         );
         this.connectionRegistry = new ConnectionRegistryImpl(
-            this.config.connectionStrategy,
+            this.config.connectionStrategy.asyncStart,
+            this.config.connectionStrategy.reconnectMode,
```
Comment (srknzl) [inline src/HazelcastClient.ts:165]: refactored constructor to accept specific configs it needs

File: src/core/MemberVersion.ts [hunk 3]
```diff
@@ -26,4 +26,16 @@ export class MemberVersion {
...
+     * @param other other version to compare to
+     * @return true if this version equals `other`
+     */
+    equals(other: MemberVersion): boolean {
+        return this.major === other.major && this.minor === other.minor;
+    }
+
+    toString(): string {
+        return `${this.major}.${this.minor}`;
+    }
```
Comment (srknzl) [inline src/core/MemberVersion.ts:40]: these are used in memberOfLargerSameVersionGroup

File: src/invocation/ClusterService.ts [hunk 4]
```diff
@@ -77,7 +77,7 @@ export class ClusterService implements Cluster {
      * @param uuid The UUID of the member as a string.
      * @return The member that was found, or undefined if not found.
      */
-    getMember(uuid: string): MemberImpl {
+    getMember(uuid: string): MemberImpl | undefined {
```
Comment (srknzl) [inline src/invocation/ClusterService.ts:80]: side change that is unrelated

File: src/network/ConnectionManager.ts [hunk 5]
```diff
@@ -148,24 +161,15 @@ export class ConnectionRegistryImpl implements ConnectionRegistry {
...
     constructor(
-        connectionStrategy: ConnectionStrategyConfig,
-        smartRoutingEnabled: boolean,
-        loadBalancer: LoadBalancer,
-        clusterService: ClusterService
+        private readonly asyncStart: boolean,
+        private readonly reconnectMode: ReconnectMode,
+        private readonly smartRoutingEnabled: boolean,
+        private readonly loadBalancer: LoadBalancer,
+        private readonly clusterService: ClusterService
```
Comment (srknzl) [inline src/network/ConnectionManager.ts:171]: preferred to use compact construction syntax(like other classes I refactored in #797 )

File: src/network/ConnectionManager.ts [hunk 6]
```diff
@@ -188,41 +192,62 @@ export class ConnectionRegistryImpl implements ConnectionRegistry {
...
+        }
+
+        const iterator = this.activeConnections.values();
+        const next = iterator.next();
+        if (!next.done) {
+            return next.value;
+        } else {
+            return null;
+        }
+    }
```
Comment (srknzl) [inline src/network/ConnectionManager.ts:213]: reverted to old version

File: src/network/ConnectionManager.ts [hunk 7]
```diff
@@ -188,41 +192,62 @@ export class ConnectionRegistryImpl implements ConnectionRegistry {
...
+            // we might be lucky...
+            for (let i = 0; i < SQL_CONNECTION_RANDOM_ATTEMPTS; i++) {
+                const member = memberOfLargerSameVersionGroup(this.clusterService.getMembers(), null);
+
+                if (member === null) {
+                    break;
+                }
                 const connection = this.getConnection(member.uuid);
-                if (connection != null) {
+                if (connection !== undefined) {
```
Comment (srknzl) [inline src/network/ConnectionManager.ts:227]: connection can only be undefined, I chose to be more specific here

File: src/sql/SqlService.ts [hunk 8]
```diff
@@ -255,7 +255,7 @@ export class SqlServiceImpl implements SqlService {
      * @param connection
      * @returns {@link HazelcastSqlException}
      */
-    toHazelcastSqlException(err: any, connection: Connection) : HazelcastSqlException {
+    rethrow(err: any, connection: Connection): HazelcastSqlException {
```
Comment (srknzl) [inline src/sql/SqlService.ts:258]: refactored so that rethrow takes err and connection and returns a HazelcastSqlException whereas toHazelcastSqlException is just for converting an err to HazelcastSqlException

File: test/unit/network/ConnectionRegistryTest.js [hunk 9]
```diff
@@ -43,33 +43,14 @@ describe('ConnectionRegistryTest', function () {
...
-
         it('should call load balancer\'s next() when in smart mode', function () {
             const loadBalancerStub = {};
             loadBalancerStub.next = sandbox.fake.returns(null);
             loadBalancerStub.nextDataMember = sandbox.spy();
 
             const connectionRegistry = new ConnectionRegistryImpl(
-                new ConnectionStrategyConfigImpl(),
+                false,
+                ReconnectMode.ON,
```
Comment (srknzl) [inline test/unit/network/ConnectionRegistryTest.js:53]: contructor change affected tests

File: test/unit/sql/SqlResult.js [hunk 10]
```diff
@@ -231,6 +231,7 @@ describe('SqlResultTest', function () {
         beforeEach(function () {
             fakeSqlService = {
                 toHazelcastSqlException: sandbox.fake((err) => new HazelcastSqlException(null, 1, '', err)),
+                rethrow: sandbox.fake((err) => new HazelcastSqlException(null, 1, '', err)),
```
Comment (srknzl) [inline test/unit/sql/SqlResult.js:234]: rethrow used in the sql service we need to add it too

File: test/unit/network/ConnectionRegistryTest.js [hunk 11]
```diff
@@ -167,102 +132,17 @@ describe('ConnectionRegistryTest', function () {
...
-                const dataMemberConnection = connectionRegistry.getRandomConnection(true);
-
-                expect(connection).to.be.equal(firstConnection);
-                expect(otherConnection).to.be.equal(firstConnection);
-                expect(dataMemberConnection).to.be.equal(null);
-
-                expect(loadBalancerStub.next.called).to.be.false;
-                expect(loadBalancerStub.nextDataMember.called).to.be.false;
-            }
-        );
```
Comment (srknzl) [inline test/unit/network/ConnectionRegistryTest.js:256]: deleted tests that are related to old sql connection selection method

File: src/sql/SqlService.ts [hunk 12]
```diff
@@ -265,31 +265,43 @@ export class SqlServiceImpl implements SqlService {
...
+            originatingMemberId = err.originatingMemberId;
+        } else {
+            originatingMemberId = this.connectionManager.getClientUuid();
+        }
+        return new HazelcastSqlException(
+            originatingMemberId, SqlErrorCode.GENERIC, err.message, err
+        );
+    }
+
+
```
Comment (mdumandag) [inline src/sql/SqlService.ts:287]: unnecessary newline

File: src/util/Util.ts [hunk 13]
```diff
@@ -286,3 +288,85 @@ export function timedPromise<T>(wrapped: Promise<T>, timeout: number, err?: Erro
 
     return deferred.promise;
 }
+
+/**
+ * Finds a larger same-version group of data members from a collection of
+ * members and, if `localMember` is from that group, return that.
+ * Otherwise return a random member from the group. If the same-version
+ * groups have the same size, return a member from the newer group
+ * (preferably the local one).
```
Comment (mdumandag) [inline src/util/Util.ts:297]: local member does not make sense for a client, so we might update the documentation

File: src/util/Util.ts [hunk 14]
```diff
@@ -286,3 +288,85 @@ export function timedPromise<T>(wrapped: Promise<T>, timeout: number, err?: Erro
...
+ * @param members list of all members
+ * @param localMember the local member, null for client instance
+ * @throws IllegalStateError If there are more than 2 distinct member versions found
+ * @return the chosen member or null, if no data member is found
+ */
+export function memberOfLargerSameVersionGroup(members: MemberImpl[], localMember: MemberImpl | null): MemberImpl | null {
+    // The members should have at most 2 different version (ignoring the patch version).
+    // Find a random member from the larger same-version group.
+
+    let version0: MemberVersion | null = null, version1: MemberVersion | null = null;
```
Comment (mdumandag) [inline src/util/Util.ts:310]: Can we define the variables on separate lines, as I think this one is harder to read

File: src/util/Util.ts [hunk 15]
```diff
@@ -286,3 +288,85 @@ export function timedPromise<T>(wrapped: Promise<T>, timeout: number, err?: Erro
...
+    for (const m of members) {
+        if (m.liteMember) {
+            continue;
+        }
+
+        const version = m.version;
+
+        let currentCount;
+
+        if (!version0 || version0.equals(version)) {
```
Comment (mdumandag) [inline src/util/Util.ts:324]: Is this check correct? I see that we have `version0 == null || ...` in Java

File: test/unit/network/ConnectionRegistryTest.js [hunk 16]
```diff
@@ -304,13 +182,12 @@ describe('ConnectionRegistryTest', function () {
...
-                    connectionStrategyConfig,
                     false,
-                    new RoundRobinLB()
+                    ReconnectMode.ASYNC,
+                    false,
+                    new RoundRobinLB(),
+                    {}
                 );
 
                 connectionRegistry.setConnectionState(connectionState.INITIALIZED_ON_CLUSTER);
```
Comment (mdumandag) [inline test/unit/network/ConnectionRegistryTest.js:193]: Can you add some tests for the new function `getConnectionForSql`, testing its behavior?

File: test/integration/backward_compatible/sql/ExecuteTest.js [hunk 17]
```diff
@@ -450,49 +450,11 @@ describe('SqlExecuteTest', function () {
         });
     });
     describe('errors/invalid usage', function () {
-        const LITE_MEMBER_CONFIG = `
```
Comment (mdumandag) [inline test/integration/backward_compatible/sql/ExecuteTest.js:453]: Instead of deleting them, we can set the maximum version of the client that this test would run to 4.2 so that we test the behavior for it.

Also, can you add a test with a cluster that has lite and data members, do some queries and make sure that it works for Node.js client as well?

File: src/BuildInfo.ts [hunk 18]
```diff
@@ -13,7 +13,8 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-/** @ignore *//** */
+/** @ignore */
```
[the remark this answers is not in the document]
Comment (srknzl) [reply src/BuildInfo.ts:16]: I think /** */ part is unnecessary but let's revert to be consistent

File: src/core/LoadBalancer.ts [hunk 19]
```diff
@@ -53,13 +53,15 @@ export interface LoadBalancer {
      * Returns the next data member to route to.
      *
      * @return Returns the next data member or `null` if no data member is available
+     * @deprecated Since 5.0, the method is unused
```
[the remark this answers is not in the document]
Comment (srknzl) [reply src/core/LoadBalancer.ts:56]: I did what java side did

File: src/core/MemberVersion.ts [hunk 20]
```diff
@@ -26,4 +26,16 @@ export class MemberVersion {
         this.minor = minor;
         this.patch = patch;
     }
+
+    /**
+     * @param other other version to compare to
+     * @return true if this version equals `other`
+     */
+    equals(other: MemberVersion): boolean {
+        return this.major === other.major && this.minor === other.minor;
```
[the remark this answers is not in the document]
Comment (srknzl) [reply src/core/MemberVersion.ts:35]: Again, took from java side, member version comparison there is like this

File: src/core/MemberVersion.ts [hunk 21]
```diff
@@ -26,4 +26,16 @@ export class MemberVersion {
...
+    }
+
+    toString(): string {
+        return `${this.major}.${this.minor}`;
```
[the remark this answers is not in the document]
Comment (srknzl) [reply src/core/MemberVersion.ts:39]: Java side copied implementation

File: src/network/ConnectionManager.ts [hunk 22]
```diff
@@ -188,41 +192,62 @@ export class ConnectionRegistryImpl implements ConnectionRegistry {
...
+    }
 
-            if (member !== null) {
+    getConnectionForSql(): Connection | null {
+        if (this.smartRoutingEnabled) {
+            // There might be a race - the chosen member might be just connected or disconnected - try a
+            // couple of times, the memberOfLargerSameVersionGroup returns a random connection,
+            // we might be lucky...
+            for (let i = 0; i < SQL_CONNECTION_RANDOM_ATTEMPTS; i++) {
+                const member = memberOfLargerSameVersionGroup(this.clusterService.getMembers(), null);
```
[the remark this answers is not in the document]
Comment (srknzl) [reply src/network/ConnectionManager.ts:221]: Yeah right, probably java side has more usages of this function. we don't need the second param

File: src/core/MemberVersion.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (mdumandag) [reply src/core/MemberVersion.ts:35]: Not MemberVersion, Version check is like this.

File: src/core/MemberVersion.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (mdumandag) [reply src/core/MemberVersion.ts:39]: Same, its like this for Version, as these are the only things it knows about. MemberVersion is a different class

File: src/util/Util.ts (hunk 15 above)
Comment (srknzl) [reply src/util/Util.ts:324]: I think it is the same behaviour, but checking against null is better, changing..

File: src/core/MemberVersion.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply src/core/MemberVersion.ts:39]: Ok, then I will change the implementation but we still need methods that work with (major,minor) pair. 

What about  

equals(other: MemberVersion, ignorePatchVersion: boolean)
toString(ignorePatchVersion: boolean)

[the remark this answers is not in the document]
Comment (mdumandag) [reply src/core/MemberVersion.ts:39]: That might do it. Another idea might be writing a utility method somewhere else and using it.
I am OK with both of them.

File: test/integration/backward_compatible/sql/ExecuteTest.js (hunk 17 above)
Comment (srknzl) [reply test/integration/backward_compatible/sql/ExecuteTest.js:453]: I realized that the test is valid for 5.0 and 4.x. I just deleted detailed error code and originatedMemberId checks to make it pass for both. Is it ok like this or should I add conditional checks? I think it is ok like this.

I added the test you mentioned

File: test/unit/network/ConnectionRegistryTest.js (hunk 16 above)
Comment (srknzl) [reply test/unit/network/ConnectionRegistryTest.js:195]: Added some unit tests

File: src/util/Util.ts [hunk 23]
```diff
@@ -286,3 +288,73 @@ export function timedPromise<T>(wrapped: Promise<T>, timeout: number, err?: Erro
...
+
+    for (const m of members) {
+        if (m.liteMember) {
+            continue;
+        }
+
+        const version = m.version;
+
+        if (version0 === null || version0.equals(version, true)) {
+            version0 = version;
```
Comment (mdumandag) [inline src/util/Util.ts:321]: you forget to increase counts

Comment (srknzl) [reply src/util/Util.ts:321]: You are right fixing thanks

File: test/integration/backward_compatible/sql/ExecuteTest.js [hunk 24]
```diff
@@ -450,22 +490,13 @@ describe('SqlExecuteTest', function () {
...
-        `;
-
         afterEach(async function () {
             await RC.terminateCluster(cluster.id);
             await client.shutdown();
         });
 
         it('should return an error if sql query sent to lite member', async function () {
-            cluster = await RC.createCluster(null, LITE_MEMBER_CONFIG);
+            cluster = await RC.createClusterKeepClusterName(null, LITE_MEMBER_CONFIG);
```
Comment (mdumandag) [inline test/integration/backward_compatible/sql/ExecuteTest.js:499]: Probably leftover change

**Review by mdumandag — APPROVED** [review]
